testkit/src/main/java/org/apache/calcite/sql/test/SqlTests.java

```diff
@@ -265,6 +265,20 @@ public static void checkEx(@Nullable Throwable ex,
       }
     }
 
+    // Search for an IllegalStateException somewhere in the stack.
+    // These are thrown by the enumerable implementors when evaluating
+    // an expression that produces an error.
+    IllegalStateException ise = null;
+    for (Throwable x = ex; x != null; x = x.getCause()) {
+      if (x instanceof IllegalStateException) {
+        ise = (IllegalStateException) x;
+        break;
+      }
+      if (x.getCause() == x) {
```

Comment:
Is this just to prevent SO?

Reply:
I have seen this pattern used elsewhere in this function, so I have followed it. I am not sure whether a Throwable can have itself as a cause. It looks like the Java APIs make that possible, since you can create a Throwable and assign a cause to it later using initCause.